src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -88,33 +115,75 @@ private void ReflectionReadMembers(object obj, XmlReaderDelegator xmlReader, Xml
                 {
                     index = context.GetMemberIndex(xmlReader, memberNames, memberNamespaces, memberIndex, null);
                 }
-                ReflectionReadMember(obj, index, xmlReader, context, memberNames, memberNamespaces);
-                memberIndex = index;
-                requiredIndex = index + 1;
+
+                // GetMemberIndex returns memberNames.Length if member not found
+                if (index < members.Length)
+                {
+                    ReflectionReadMember(obj, index, xmlReader, context, members);
+                    memberIndex = index;
+                    requiredIndex = index + 1;
+                }
             }
         }
 
-        private int ReflectionReadMember(object obj, int memberIndex, XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString[] memberNames, XmlDictionaryString[] memberNamespaces)
+        // Put all members of the contract (including types from base contract) into 'members'.
+        private int ReflectionGetMembers(ClassDataContract classContract, DataMember[] members)
         {
-            int memberCount = 0;
+            int memberCount = (classContract.BaseContract == null) ? 0 : ReflectionGetMembers(classContract.BaseContract, members);
+            int childElementIndex = memberCount;
+            for (int i = 0; i < classContract.Members.Count; i++, memberCount++)
+            {
+                members[childElementIndex + i] = classContract.Members[i];
```

Comment:
How do we know members[] is big enough?  If is created outside this method based on MemberNames.Length, so it is not immediately clear why this is big enough for what appears to be a recursive method to fill in the members of the inheritance hierarchy.  Perhaps a Contract.Assert() belongs here to ensure we don't index past the end.